mail_tracking_mailgun/models/mail_tracking_email.py

```diff
@@ -0,0 +1,196 @@
+from openerp import models, api, fields
+
+import logging
+_logger = logging.getLogger(__name__)
+
+
+class MailTrackingEmail(models.Model):
+    _inherit = "mail.tracking.email"
+
+    def _country_search(self, country_code):
+        country = False
+        if country_code:
+            country = self.env['res.country'].search([
+                ('code', '=ilike', country_code),
+            ])
+        if country:
+            return country.id
+        return False
+
+    def _mailgun_mandatory_fields(self):
+        return ('event', 'timestamp', 'token', 'signature',
+                'tracking_email_id', 'odoo_db')
+
+    def _mailgun_event_type_mapping(self):
+        return {
+            # Mailgun event type: tracking event type
+            'delivered': 'delivered',
+            'opened': 'open',
+            'clicked': 'click',
+            'unsubscribed': 'unsub',
+            'complained': 'spam',
+            'bounced': 'hard_bounce',
+            'dropped': 'reject',
+        }
+
+    def _mailgun_supported_event_types(self):
+        return self._mailgun_event_type_mapping().keys()
+
+    def _mailgun_event_type_verify(self, event):
+        event = event or {}
+        mailgun_event_type = event.get('event')
+        if mailgun_event_type not in self._mailgun_supported_event_types():
+            _logger.info("Mailgun: event type '%s' not supported",
+                         mailgun_event_type)
+            return False
+        # OK, event type is valid
+        return True
+
+    def _mailgun_signature(self, api_key, timestamp, token):
+        return hmac.new(
+            key=str(api_key),
+            msg='{}{}'.format(str(timestamp), str(token)),
+            digestmod=hashlib.sha256).hexdigest()
+
+    def _mailgun_signature_verify(self, event):
+        event = event or {}
+        api_key = self.env['ir.config_parameter'].get_param('mailgun.apikey')
+        if not api_key:
+            _logger.info("No Mailgun api key configured. "
+                         "Please add 'mailgun.apikey' to System parameters "
+                         "to enable Mailgun authentication webhoook requests. "
+                         "More info at: "
+                         "https://documentation.mailgun.com/user_manual.html"
+                         "#webhooks")
+        else:
+            timestamp = event.get('timestamp')
+            token = event.get('token')
+            signature = event.get('signature')
+            event_digest = self._mailgun_signature(api_key, timestamp, token)
+            if signature != event_digest:
+                _logger.info("Mailgun: Invalid signature '%s' != '%s'",
+                             signature, event_digest)
+                return False
+        # OK, signature is valid
+        return True
+
+    def _db_verify(self, event):
+        event = event or {}
+        odoo_db = event.get('odoo_db')
+        current_db = getattr(threading.currentThread(), 'dbname', None)
```

Comment:
The same question as above: shouldn't we get the DB name from `self.env.cr.dbname` or similar?